File: typescript/Behavioral/19-State/Context/JobCard.ts
import type { IJobState } from "../State/IJobState";
import { DraftState } from "../State/DraftState";

export class JobCard {
  readonly jobId: string = Math.random().toString(36).slice(2, 10);

  assignee: string | undefined;
  pauseReason: string | undefined;
  startedAt: Date | undefined;
  completedAt: Date | undefined;
  closedAt: Date | undefined;

  private state: IJobState;

  constructor(
    readonly title: string,
    readonly category: string,
  ) {
    this.state = new DraftState(); // initial state
  }

  get currentState(): string {
    return this.state.name;
  }

  // Called by state classes to move to the next state
  transitionTo(newState: IJobState): void {
    console.log(`  [${this.state.name}] → [${newState.name}]`);
    this.state = newState;
  }

  // Delegate all actions to the current state
  assign(handyman: string): void {
    this.state.assign(this, handyman);
  }
  start(): void {
    this.state.start(this);
  }
  pause(reason: string): void {
    this.state.pause(this, reason);
  }
  resume(): void {
    this.state.resume(this);
  }
  complete(): void {
    this.state.complete(this);
  }
  approve(): void {
    this.state.approve(this);
  }

  toString(): string {
    return (
      `Job #${this.jobId} '${this.title}' [${this.category}] — State: ${this.state.name}` +
      (this.assignee ? ` | ${this.assignee}` : "")
    );
  }
}

File: csharp/Behavioral/19-State/Context/JobCard.cs
namespace State;

public class JobCard
{
    public string JobId { get; } = Guid.NewGuid().ToString("N")[..8];

    public string Title { get; }
    public string Category { get; }

    public string? Assignee { get; set; }
    public string? PauseReason { get; set; }
    public DateTime? StartedAt { get; set; }
    public DateTime? CompletedAt { get; set; }
    public DateTime? ClosedAt { get; set; }

    private IJobState _state;

    public JobCard(string title, string category)
    {
        Title = title;
        Category = category;
        _state = new DraftState(); // initial state
    }

    public string CurrentState => _state.Name;

    // Called by state classes to move to the next state
    public void TransitionTo(IJobState newState)
    {
        Console.WriteLine($"  [{_state.Name}] → [{newState.Name}]");
        _state = newState;
    }

    // Delegate all actions to the current state
    public void Assign(string handyman) => _state.Assign(this, handyman);
    public void Start()                 => _state.Start(this);
    public void Pause(string reason)    => _state.Pause(this, reason);
    public void Resume()                => _state.Resume(this);
    public void Complete()              => _state.Complete(this);
    public void Approve()               => _state.Approve(this);

    public override string ToString() =>
        $"Job #{JobId} '{Title}' [{Category}] — State: {_state.Name}" +
        (Assignee is not null ? $" | {Assignee}" : "");
}
